src/EffectiveCSharp.Analyzers/PreferDeclarationInitializersToAssignmentStatementsAnalyzer.cs

```diff
@@ -0,0 +1,656 @@
+        helpLinkUri: HelpLinkUri);
+
+    private static readonly DiagnosticDescriptor RuleExceptionShouldNotInitializeToNullOrZero = new(
+        id: DiagnosticIds.PreferDeclarationInitializersExceptNullOrZero,
+        title: "Should not initialize to null or zero",
+        messageFormat: "Do not initialize to null or zero as these already occur by default",
+        category: "Maintainability",
+        defaultSeverity: DiagnosticSeverity.Info,
+        isEnabledByDefault: true,
+        description: "Field initialization to null or zero is redundant and should be avoided.",
+        helpLinkUri: HelpLinkUri);
+
+    private static readonly DiagnosticDescriptor RuleExceptionShouldNotInitializeInDeclaration = new(
+        id: DiagnosticIds.PreferDeclarationInitializersExceptWhenVaryingInitializations,
+        title: "Should not initialize in declaration due to diverging initializations in constructors",
+        messageFormat: "Do not initialize a field in its declaration if you have diverging initializations in constructors. This is to prevent unnecessary allocations.",
+        category: "Maintainability",
+        defaultSeverity: DiagnosticSeverity.Info,
+        isEnabledByDefault: true,
+        description: "Field initialization should not occur when there are diverging initializations in constructos. This is to prevent unnecessary allocations.",
+        helpLinkUri: HelpLinkUri);
+
+    private static readonly DiagnosticDescriptor RuleShouldInitializeInDeclarationWhenNoInitializationPresent = new(
+        id: DiagnosticIds.PreferDeclarationInitializersWhenNoInitializationPresent,
+        title: "Should initialize in declaration when no initialization present",
+        messageFormat: "Initialize the field in its declaration when no distint initializations will occur in constructors",
+        category: "Maintainability",
+        defaultSeverity: DiagnosticSeverity.Info,
+        isEnabledByDefault: true,
+        description: "Field initialization should occur in the declaration unless there are diverging initializations in constructors, or the field is a value type or nullable being initialized to the default.",
+        helpLinkUri: HelpLinkUri);
+
+    /// <inheritdoc />
+    public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(GeneralRule, RuleExceptionShouldNotInitializeToNullOrZero, RuleExceptionShouldNotInitializeInDeclaration, RuleShouldInitializeInDeclarationWhenNoInitializationPresent);
+
+    /// <inheritdoc />
+    public override void Initialize(AnalysisContext context)
+    {
+        context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+        context.EnableConcurrentExecution();
+        context.RegisterSyntaxNodeAction(AnalyzeNode, SyntaxKind.ClassDeclaration);
+    }
+
+    private static void AnalyzeNode(SyntaxNodeAnalysisContext context)
```

Comment:
In general, you want to minimize object creation in hot paths. `FieldInitializationInfo` sticks out to me

Reply:
Agreed, but I need to cache some of that information somewhere, hence why the class is valuable. If there's a much more memory efficient way to cache the expression nodes where the initializations occur, I am happy to try it out